### PR TITLE
Provide an option to select unique_id generator

### PR DESCRIPTION
**Pull request by kbialek**
# What does this implement/fix? 

Provides fully backwards compatible option to specify alternative unique_id generator.
The existing generator is retained, and used as the default one.
Additionally a new, mac-address based generator is implemented.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** 
* closes https://github.com/esphome/esphome/pull/2280
* https://github.com/esphome/issues/issues/1637

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/1610

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:
```yaml
# Default
mqtt:
  discovery_unique_id_generator: legacy

# mac-address based
mqtt:
  discovery_unique_id_generator: mac
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — CHANGES_REQUESTED** [review]
Please change this to use an enum, not a string

File: esphome/components/mqtt/mqtt_client.cpp [hunk 1]
```diff
@@ -535,8 +535,10 @@ void MQTTClientComponent::set_birth_message(MQTTMessage &&message) {
 
 void MQTTClientComponent::set_shutdown_message(MQTTMessage &&message) { this->shutdown_message_ = std::move(message); }
 
-void MQTTClientComponent::set_discovery_info(std::string &&prefix, bool retain, bool clean) {
+void MQTTClientComponent::set_discovery_info(std::string &&prefix, MQTTDiscoveryUniqueIdGenerator &&unique_id_generator,
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_client.cpp]:
```suggestion
void MQTTClientComponent::set_discovery_info(std::string &&prefix, MQTTDiscoveryUniqueIdGenerator unique_id_generator,
```

File: esphome/components/mqtt/mqtt_client.h [hunk 2]
```diff
@@ -98,9 +105,11 @@ class MQTTClientComponent : public Component {
    *
    * See <a href="https://www.home-assistant.io/docs/mqtt/discovery/">MQTT Discovery</a>.
    * @param prefix The Home Assistant discovery prefix.
+   * @param unique_id_generator Controls how UniqueId is generated.
    * @param retain Whether to retain discovery messages.
    */
-  void set_discovery_info(std::string &&prefix, bool retain, bool clean = false);
+  void set_discovery_info(std::string &&prefix, MQTTDiscoveryUniqueIdGenerator &&unique_id_generator, bool retain,
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_client.h]:
```suggestion
  void set_discovery_info(std::string &&prefix, MQTTDiscoveryUniqueIdGenerator unique_id_generator, bool retain,
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Comment by oxan** [comment]
If we're gonna change the unique_id generation, I'd like to get it right, specifically esphome/issues#2028 (non-Latin characters are stripped in the `default_object_id`). I think allowing non-ASCII characters into the unique id is asking for problems, so it's probably best if we hash the name and use that.

**Comment by kbialek** [comment]
@oxan This goal of this PR is to add an option to generate more distinct `unique_id`s, by using `mac-address` where `ESP` string was used so far. It does not support setting any free text as the `unique_id`.

**Comment by oxan** [comment]
I understand that, but changing the `unique_id` algorithm is a massive pain, so if we do that, I'd like to fix all issues with it at once instead of having to do it again later on. Both the old and your proposed new algorithm include user input in the form of the entity name in the unique_id through the `default_object_id`.

**Comment by kbialek** [comment]
As far as I understood from the comments in the linked PR/issue, changing the existing `unique_id` generation is not an option. This means, that the problem you linked cannot be fully resolved. 
@oxan What the new generator should use instead of `default_object_id` in your opinion?

**Comment by oxan** [comment]
I think a hex-encoded FNV-1 hash of the name would be fine, though maybe we should use a 64-bit or 128-bit hash instead.

**Comment by kbialek** [comment]
I found `fnv1_hash` function in the codebase. Is that what you mean?

**Comment by oxan** [comment]
Yes.

**Comment by kbialek** [comment]
@oxan I've implemented the change. Now the `unique_id` looks like this `807d3a3c2190-switch-d3a92fe4` for example. Please review. The documentation is not yet updated.

File: esphome/components/mqtt/mqtt_component.cpp [hunk 3]
```diff
@@ -114,9 +114,17 @@ bool MQTTComponent::send_discovery_() {
         if (!unique_id.empty()) {
           root[MQTT_UNIQUE_ID] = unique_id;
         } else {
-          // default to almost-unique ID. It's a hack but the only way to get that
-          // gorgeous device registry view.
-          root[MQTT_UNIQUE_ID] = "ESP" + this->component_type() + this->get_default_object_id_();
+          const MQTTDiscoveryInfo &discovery_info = global_mqtt_client->get_discovery_info();
+          if (discovery_info.unique_id_generator == MQTT_MAC_ADDRESS_UNIQUE_ID_GENERATOR) {
+            char object_id_hash[9];
+            sprintf(object_id_hash, "%08x", fnv1_hash(this->get_default_object_id_()));
```
Comment (oxan) [inline esphome/components/mqtt/mqtt_component.cpp]: This doesn't work, as `get_default_object_id()` already strips non-Latin characters. We should hash the raw name instead:
```suggestion
            sprintf(object_id_hash, "%08x", fnv1_hash(this->friendly_name()));
```

**Review by oxan — CHANGES_REQUESTED** [review]

**Review by oxan — APPROVED** [review]
LGTM, though I'd like a second opinion on whether we maybe should use a 64-bit or 128-bit hash instead.

**Comment by kbialek** [comment]
Documentation PR updated https://github.com/esphome/esphome-docs/pull/1610/commits/1faf373c85fffc061d45146668847e0bd03544d0

**Review by jesserockz — APPROVED** [review]
> LGTM, though I'd like a second opinion on whether we maybe should use a 64-bit or 128-bit hash instead.

I don't think we need to go further with the hash. Just making it longer is not going to make it more unique.

**Comment by oxan** [comment]
> Just making it longer is not going to make it more unique.

Why not? There's (usually) more than 32 bits of entropy in the names, so longer hashes should've less collisions.

**Comment by jesserockz** [comment]
Yeah, I just mean in this case, do you really think you will get a clash of the `friendly_name` hash within a single ESP device?

I am not against changing it btw, just a thought.

**Comment by oxan** [comment]
Yeah fair enough, that's probably quite unlikely.

**Comment by oxan** [comment]
Thanks!

**Pull request merged**